### ISSUE TITLE
Cannot display step-based progress with format alone

### ISSUE THREAD

**Issue by XedinUnknown** [wp-cli/wp-cli#5125]
## The Problem
Sometimes, it could be necessary to display not the percentage of the progress, but how many steps are completed out of a total. For example, something like this:

```
5/10 [==============================>                              ] 0:00 / 0:00
```

Instead of something like this:

```
50% [==============================>                              ] 0:00 / 0:00
```

Currently, this is not possible without overriding the [`Bar`][] class.

## Possible Cause
1. The constructor of `Bar`, declared in [`Progress`], [does not accept a format parameter][1].
2. The placeholders [passed][2] to `Streams::render()` do not include the [`_current`][`Notify#_current`] and [`_total`][`Progress#_total`] values.

## Suggested Solution
1. Accept a format argument in the constructor.
2. Pass the necessary values of the respective members for `_current` and `_total`.

## Remarks
What about other formats, like [timing][`Bar#_formatTiming`]? Should they also be accepted as parameters?

[`Bar`]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/progress/Bar.php
[`Bar#_formatTiming`]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/progress/Bar.php#L31
[`Notify#_current`]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/Notify.php#L27
[`Progress#_total`]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/Progress.php#L23
[`Progress`]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/Progress.php

[1]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/Progress.php#L33
[2]: https://github.com/wp-cli/php-cli-tools/blob/master/lib/cli/progress/Bar.php#L52

**Comment by danielbachhuber** [comment]
> * Accept a format argument in the constructor.
> 
> * Pass the necessary values of the respective members for `_current` and `_total`.

This seems reasonable. Want to put together a pull request for it?

**Comment by XedinUnknown** [comment]
@danielbachhuber, yep. But then, like I write in the Remarks section, should we accept an argument for all of the formats? Because there are more.

**Comment by danielbachhuber** [comment]
> should we accept an argument for all of the formats? Because there are more.

I think we should only start with those that we need. We can always iterate later.

**Comment by XedinUnknown** [comment]
What's "those we need" though? I specifically need to edit the format of the message. But that's just me. Someone else may need to format the timing, the bar, or the whole thing.

**Comment by schlessera** [comment]
I actually think it would make more sense to not add to the number of arguments, but just provide a different implementation in your own code if you have special needs. I don't see anything inherently wrong with replacing the `Bar` class with your own `StepBasedBar` implementation.

Is there a specific reason why you would want to avoid that?

**Comment by XedinUnknown** [comment]
Well, was just looking for a way to do this, and noticed that currently the `Bar` class decides the format, which breaks SOLID. So, I thought that it would be useful to make it configurable, and improve the architecture.

In order to not make the constructor too complicated, here's how I would solve this:

```php
interface ProgressBarDesignInterface
{
	public function getFractionFormat(): string;

	public function getMessageFormat(): string;

	public function getTimingFormat(): string;

	public function getBarFormat(): string;
}

// Elseweyr...

$bar = new Bar((new class() implements ProgressBarDesignInterface {
	public function getFractionFormat(): string
	{
		return '=';
	}

	public function getMessageFormat(): string
	{
		return '{:msg}  {:percent}% [';
	}

	public function getTimingFormat(): string
	{
		return '] {:elapsed} / {:estimated}';
	}

	public function getBarFormat(): string
	{
		return '{:msg}{:bar}{:timing}';
	}
}));
```

Of course, the passing of values to the format function needs to be fixed either way.

A variation of this could be the `ProgressBarDesignInterface` methods accepting value maps, and doing the formatting. But that would make the implementation more complex.

While fixing the missing format function parameters is something that I can solve right here, the configurability of the `Bar` itself feels like another task for `3.x`. Is there a discussion on this topic somewhere?

**Comment by crstauf** [comment]
FYI, I accomplish this by doing the following:

```php
$rows = range( 1, 5 );
$progress_bar = WP_CLI\Utils\make_progress_bar( sprintf( 'Processing %d', count( $rows ) ), count( $rows );

foreach ( $rows as $row ) {
    $progress_bar->tick( 1, sprintf( 'Processed %s/%s.', trim( $progress_bar->current() ), $progress_bar->total() ) );
}

$progress_bar->finish();
```
